woocommerce/woocommerce-gutenberg-products-block
# Rest API batching support

This PR implements batching in the Store API, and a dataloader class in the client to batch requests within a certain interval.

Batching works for POST, PUT, DELETE API requests. GET requests are not affected.

As well as adding the batching, [this commit](https://github.com/woocommerce/woocommerce-gutenberg-products-block/commit/d0b901b06174aa514efaf23bec1482db13f6b276) creates a control to handle fragment updates and adds a timeout. This prevents multiple fragment updates after a batch of events have ran.

NOTE: I am requested feedback/code review, but some things are left to do before it can be merged:

- [x] Check what happens if nonce changes on items in the batch.
- [x] Batch is only supported on 5.6+ so we need the client to account for this also.
- [x] Reduce batch timeout from 3000 - I have left this high to make testing easier.

Fixes #2094

### How to test the changes in this Pull Request:

1. Go to the products page.
2. Add to items to the cart really quickly. If you focus a button with the keyboard you can hit enter and click another button.
3. Check the network request for a "batch" API call. You'll see 2 responses in an array.
4. Smoke test cart page functionality/cart/checkout. All should be unaffected.
